### PR TITLE
Fix mac token cache thread safety

### PR DESCRIPTION
**Pull request by antrix1989**
## Proposed changes

Describe what this PR is trying to do.

## Type of change

- [ ] Feature work
- [ ] Bug fix
- [ ] Documentation
- [ ] Engineering change
- [ ] Test
- [ ] Logging/Telemetry

## Risk

- [ ] High – Errors could cause MAJOR regression of many scenarios. (Example: new large features or high level infrastructure changes)
- [ ] Medium – Errors could cause regression of 1 or more scenarios. (Example: somewhat complex bug fixes, small new features)
- [x] Small – No issues are expected. (Example: Very small bug fixes, string changes, or configuration settings changes)

### CODE REVIEW

**Review by kaisong1990 — APPROVED** [review]
Code looks good to me. Just curious, what makes the decision of blocking the thread instead of making another deep copy?

File: IdentityCore/src/cache/MSIDMacTokenCache.m [hunk 1]
```diff
@@ -488,8 +485,8 @@ - (BOOL)setItemImpl:(MSIDCredentialCacheItem *)item
     }
     
     __block NSDictionary *tokens;
-    dispatch_sync(self.synchronizationQueue, ^{
-        tokens = [[self.cache objectForKey:@"tokens"] copy];
+    dispatch_barrier_sync(self.synchronizationQueue, ^{
```
Comment (githubursul) [inline IdentityCore/src/cache/MSIDMacTokenCache.m:488]: I am not sure how these two former dispatch_sync blocks were racing. self.cache seems to be accessed only for read. Am I missing anything?

**Review by githubursul — COMMENTED** [review]
Are you confident this fixes the issue?

File: IdentityCore/src/cache/MSIDMacTokenCache.m (hunk 1 above)
Comment (githubursul) [reply IdentityCore/src/cache/MSIDMacTokenCache.m:488]: Understood. Either lock (i.e. barrier) or deep copy is required.

**Review by githubursul — APPROVED** [review]
:shipit:

**Comment by antrix1989** [comment]
> Code looks good to me. Just curious, what makes the decision of blocking the thread instead of making another deep copy?

There is no much benefit doing a copy performance-wise, the full deep copy is created in the same method with NSKeyedArchiver.

**Pull request merged**